### PR TITLE
Fix: The termination grace period is set to maximum explicitly defined values, otherwise, it is set to null.

### LINKED ISSUE #16200: Workspace stop time increased sufficiently in the 7.9.0 version
### Describe the bug
Workspace stop time increased sufficiently after the 7.9.0 update from 2 to 40 seconds

image

### Che version

 - [x] latest
 - [x] nightly
 - [ ] other: please specify

### Steps to reproduce

- Start workspace
- Wait until started
- Stop Workspace
- Error: workspace will be marked as stopped only after 40 seconds

### Additional context
The current suspect is  `che.infra.kubernetes.pod.termination_grace_period_sec=0` that might not be set in the pod spec. By default `terminationGracePeriodSeconds` is `30` on k8s

https://github.com/eclipse/che/blob/master/assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties#L405

### PR DESCRIPTION
**Pull request by skabashnyuk**
### What does this PR do?
The termination grace period is set to maximum explicitly defined values, otherwise, it is set to null.
Note, that this is related to the loop that is happened inside of PodMerger. Before it, there is another loop other all pods that set https://github.com/eclipse/che/blob/master/infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/PodTerminationGracePeriodProvisioner.java#L51
Basically what is happening in the bubble sort inside    https://github.com/eclipse/che/blob/master/infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/environment/PodMerger.java#L72 this loop. I believe that because of PodTerminationGracePeriodProvisioner at the end erminationGracePeriod should never be null. 

### What issues does this PR fix or reference?
Fixes: https://github.com/eclipse/che/issues/16200

#### Release Notes
n/a

#### Docs PR
n/a

### CODE REVIEW

File: infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/environment/PodMergerTest.java [hunk 1]
```diff
@@ -237,6 +237,57 @@ public void shouldNotAddImagePullPolicyTwice() throws Exception {
...
+    PodData podData1 = new PodData(podSpec1, new ObjectMetaBuilder().build());
+    PodSpec podSpec2 = new PodSpecBuilder().withTerminationGracePeriodSeconds(25L).build();
+    PodData podData2 = new PodData(podSpec2, new ObjectMetaBuilder().build());
+
+    // when
+    Deployment merged = podMerger.merge(Arrays.asList(podData1, podData2));
+
+    // then
+    PodTemplateSpec podTemplate = merged.getSpec().getTemplate();
+    assertEquals(podTemplate.getSpec().getTerminationGracePeriodSeconds(), (Long) 42L);
```
Comment (sparkoo) [inline infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/environment/PodMergerTest.java]: why this assert? shouldn't it throw an exception during the merge on Line 284 and end there?

File: infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/environment/PodMergerTest.java [hunk 2]
```diff
@@ -237,6 +237,57 @@ public void shouldNotAddImagePullPolicyTwice() throws Exception {
...
+    // then
+    PodTemplateSpec podTemplate = merged.getSpec().getTemplate();
+    assertEquals(podTemplate.getSpec().getTerminationGracePeriodSeconds(), (Long) 24L);
+  }
+
+  @Test(
+      expectedExceptions = ValidationException.class,
+      expectedExceptionsMessageRegExp =
+          "Cannot merge pods with a different configuration of the termination grace period: 24, 25")
+  public void shouldFailIfTerminationGracePeriodSharedBDiffersInPods() throws Exception {
```
Comment (sparkoo) [inline infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/environment/PodMergerTest.java]: typo in method name `...SharedBDiffers...`

File: infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/environment/PodMerger.java [hunk 3]
```diff
@@ -116,6 +116,11 @@ public Deployment merge(List<PodData> podsData) throws ValidationException {
         }
       }
 
+      baseSpec.setTerminationGracePeriodSeconds(
```
Comment (ibuziuk) [inline infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/environment/PodMerger.java]: IMO, we should always fall back on `terminationGracePeriodSeconds` aka baseSpec.

**Review by metlos — COMMENTED** [review]
Apart from the typos and leftovers in the tests that @sparkoo pointed out, I think we should think how the termination grade period differs in semantics from, say, a service account name.

For service account name, the requirement for it to be either undefined or the same in all pods makes sense, because we actually cannot handle multiple service accounts for a single pod.

I think the situation with the grace period is actually slightly different. The value expresses the wait time before containers in a pod are forcibly killed during pod termination. This is to give the applications in the pod time to gracefully react to shutdown.

So requiring the same grace period doesn't make sense as different applications have just different requirements. IMHO, here we should actually find the maximum grace period and apply that to the resulting merged pod. This ensures that all applications have been given at least their required grace period.

**Comment by ibuziuk** [comment]
> So requiring the same grace period doesn't make sense as different applications have just different requirements

@metlos This is a good point, but do we have such a situation in Che e.g. what will be the side-effects in using explicit `terminationGracePeriodSeconds` if set via property / env var? 

If the property is not set, of course, we can add some fancy logic, but I just not sure if it worth it :man_shrugging:

**Comment by metlos** [comment]
> > So requiring the same grace period doesn't make sense as different applications have just different requirements
> 
> @metlos This is a good point, but do we have such a situation in Che e.g. what will be the side-effects in using explicit `terminationGracePeriodSeconds` if set via property / env var?
> 
> If the property is not set, of course, we can add some fancy logic, but I just not sure if it worth it man_shrugging

This code is also used for deploying devfiles where we can have free-form user-defined kubernetes lists, so I think it makes sense to consider this.

File: infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/environment/PodMergerTest.java [hunk 4]
```diff
@@ -237,6 +237,57 @@ public void shouldNotAddImagePullPolicyTwice() throws Exception {
...
+    // when
+    Deployment merged = podMerger.merge(Arrays.asList(podData1, podData2));
+
+    // then
+    PodTemplateSpec podTemplate = merged.getSpec().getTemplate();
+    assertEquals(podTemplate.getSpec().getTerminationGracePeriodSeconds(), (Long) 24L);
+  }
+
+  @Test
+  public void shouldMergeTerminationGracePeriodSharedByPodsIfOneIsNull() throws Exception {
```
Comment (amisevsk) [inline infrastructures/kubernetes/src/test/java/org/eclipse/che/workspace/infrastructure/kubernetes/environment/PodMergerTest.java]: Maybe add test case where other termination grace period is null.

**Review by amisevsk — APPROVED** [review]
LGTM, with the other reviewers' suggestions.

**Comment by skabashnyuk** [comment]
@sparkoo @metlos @ibuziuk @amisevsk I've change algorithm can you please reevaluate implementation.

**Comment by sparkoo** [comment]
You've changed behavior from throwing an exception to setting the null if all is null. So what happens in that case ?

**Comment by skabashnyuk** [comment]
>You've changed behavior from throwing an exception to setting the null if all is null. So what happens in that case ?

From PodMerger perspective if all pods have null - > result will have null. And it would be up to cluster configuration to set this value. Afaik 30 sec by default.

From Che global perspective, all null values would be overridden  by https://github.com/eclipse/che/blob/master/infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/provision/PodTerminationGracePeriodProvisioner.java#L51

**Review by sparkoo — APPROVED** [review]

**Review by metlos — APPROVED** [review]

**Review by ibuziuk — APPROVED** [review]
:+1: would be great to have it in 7.9.1 also

**Comment by skabashnyuk** [comment]
ci-build

**Comment by skabashnyuk** [comment]
ci-test

**Comment by skabashnyuk** [comment]
[ci-test]

**Comment by skabashnyuk** [comment]
[ci-build]
